### PR TITLE
Fix inconsistent field-level validation (Fixes #204)

### LINKED ISSUE #180: [Bug] All RuleSets firing, even though only one specific in component options
**Describe the bug**
I am specifying an explicit RuleSet to include when using the FluentValidationValidator component in my component, e.g.
`<FluentValidationValidator Options="@(options => options.IncludeRuleSets(CreateUserModelValidator.ModifyUserRuleSet))" @ref="fluentValidationValidator" />`. However, all RuleSets are firing which is printing duplicate validation messages on the screen. 

**NOTE:** I am not explicitly calling Validate(), Fluent Validation seems to be checking for valid fields when those fields are updated.

**Expected behavior**
Only the specified RuleSet is executed when updating any fields.

**Screenshots**
Duplicate validation messages:
image

Redacated validator class with RuleSets:
```
RuleSet(CreateUserRuleSet, () =>
{
    ...
    When(u => u.LoginEnabled && RequiresNPI(u), () =>
    {
        RuleFor(x => x.NPI)
            ...
            .Matches(new Regex("\\d{10}"))
                .WithMessage("NPIs must be 10 digit numbers");
    });
});

RuleSet(ModifyUserRuleSet, () =>
{
  When(u => u.LoginEnabled && RequiresNPI(u), () =>
  {
      RuleFor(x => x.NPI)
          ...
          .Matches(new Regex("\\d{10}"))
              .WithMessage("NPIs must be 10 digit numbers")
  });
});

RuleSet(ManualValidateNPIRuleSet, () =>
{
    RuleFor(p => p.NPI)
        ...
        .Matches(new Regex("\\d{10}"))
            .WithMessage("NPIs must be 10 digit numbers");
});
```

**Hosting Model (is this issue happening with a certain hosting model?):**
 - Blazor Server

### LINKED ISSUE #188: Validation on childcollection fails if using the Validator property
**Setup**
Hosted Blazor WebAssembly (default template in VS)
FluentValidation 11.5.2
Blazored.FluentValidation 2.1.0

The code has three ways validate a personcollection (2persons in each collection, one collection per example)
"Assembly validation" is using separate validator classes and detected by assembly scanning.
"Validator property" is using separate validator classes, but instead of using scanning the Validator property is given an instance of the PersonCollectionValidator.
"Inline validator" is using an instance of inline validators, where the IValidator<PersonCollection> instance is given to the Validator property of then FluentValidationValidator.

**Expected**
For all examples.
If a modified input value fails validation a validation message should appear below the input.
When clicking the submit button, all fields that fails validation a validation message should appear below each input.

**Actual**
For "Assembly validation" all works as expected.
For the other two examples that are using the Validation property validation works when clicking on the submit button, but an exception is thrown when the input lost focus and an error message appears in the browser console stating "System.InvalidOperationException: Cannot validate instances of type 'PersonClass'. This validator can only validate instances of type 'PersonCollection'."
image

**Code**
(index.razor is modified)
```

@page "/"
@using System.ComponentModel.DataAnnotations;
@using Blazored.FluentValidation;
@using FluentValidation;
@using Microsoft.Extensions.Localization;
@using System.Diagnostics;

<PageTitle>Index</PageTitle>

<h1>Assembly validation</h1>
<EditForm Model=@pc1>
    <FluentValidationValidator  />
    @foreach(var person in pc1.Persons) {
        <div class="form-group">
            <label for="Name">Name</label>
            <InputText @bind-Value="@person.Name"/>
            <ValidationMessage For=@( () => person.Name) />
        </div>
        <div class="form-group">
            <label for="Age">Age</label>
            <InputNumber @bind-Value=person.Age class="form-control" id="Age" />
            <ValidationMessage For=@( () => person.Age ) />
        </div>
    }


    <input type="submit" class="btn btn-primary" value="Save" />
</EditForm>
<br/>
<h1>Validator property</h1>
<EditForm Model=@pc2>
    <FluentValidationValidator Validator="@validator" />
    @foreach (var person in pc2.Persons)
    {
        <div class="form-group">
            <label for="Name">Name</label>
            <InputText @bind-Value="@person.Name"/>
            <ValidationMessage For=@( () => person.Name) />
        </div>
        <div class="form-group">
            <label for="Age">Age</label>
            <InputNumber @bind-Value=person.Age class="form-control" id="Age" />
            <ValidationMessage For=@( () => person.Age ) />
        </div>
    }


    <input type="submit" class="btn btn-primary" value="Save" />
</EditForm>
<br/>
<h1>Inline validator</h1>
<EditForm Model=@pc3>
    <FluentValidationValidator Validator="@inlineValidator" />
    @foreach (var person in pc3.Persons)
    {
        <div class="form-group">
            <label for="Name">Name</label>
            <InputText @bind-Value="@person.Name"/>
            <ValidationMessage For=@( () => person.Name) />
        </div>
        <div class="form-group">
            <label for="Age">Age</label>
            <InputNumber @bind-Value=person.Age class="form-control" id="Age" />
            <ValidationMessage For=@( () => person.Age ) />
        </div>
    }


    <input type="submit" class="btn btn-primary" value="Save" />
</EditForm>

@code {
    PersonCollection pc1 = new PersonCollection();
    PersonCollection pc2 = new PersonCollection();
    PersonCollection pc3 = new PersonCollection();

    PersonClass Person = new PersonClass();
    IValidator<PersonCollection> validator = new PersonCollectionValidator();
    IValidator inlineValidator = CreateInlineValidator();

    private static IValidator CreateInlineValidator()
    {
        var collectionVal = new InlineValidator<PersonCollection>();
        var personVal = new InlineValidator<PersonClass>();
        personVal.RuleFor(x => x.Name).NotEmpty();
        collectionVal.RuleForEach(x => x.Persons).SetValidator(personVal);

        return collectionVal;
    }

    public class PersonCollection
    {
        public IEnumerable<PersonClass> Persons { get; set; } = new List<PersonClass>
        {
            new PersonClass { Age=1, Name="A"},
            new PersonClass { Age=2, Name="B"}
        };

        public IValidator GetValidator()
        {
            var inlineVal = new PersonCollectionValidator();
            inlineVal.RuleForEach(x => x.Persons).SetValidator(new PersonClassValidator());
            return inlineVal;
        }
    }

    public class PersonClass
    {
        public string Name { get; set; }
        public int Age { get; set; }
    }

    public class PersonCollectionValidator : AbstractValidator<PersonCollection>
    {
        public PersonCollectionValidator()
        {
            RuleForEach(x => x.Persons).SetValidator(new PersonClassValidator());
        }
    }

    public class PersonClassValidator : AbstractValidator<PersonClass>
    {
        public PersonClassValidator()
        {
            RuleFor(x => x.Name).NotEmpty();
        }
    }
}
```

### LINKED ISSUE #204: [Bug] Inconsistent model-level and field-level behaviour for sub-objects
**Describe the bug**
With an EditContext model containing one or more sub-objects, eg a `Customer` which has an `Address`, the
`FieldIdentifier.Model` for the changed field will be different to the root object when `Address` properties have changed. For example, if `Address.FirstLine` changes, `FieldIdentifier.Model` will be of type `Address`, so a registered `IValidator<Address>` validator will be used to validate at the field-level instead of `IValidator<Customer>` like at the model-level.

This means an exception is thrown when blurring out of Address fields if an `IValidator<Customer>` is provided to `<FluentValidationValidator>` via the `Validator` property.

If a Validator is not provided, and one is instead retrieved from DI / assembly scan, then any rules affecting the `Address` that are defined on the `IValidator<Customer>` validator will not be executed, leading to inconsistent behaviour between model-level and field-level validation.

Also, when validating at the field-level, any configuration provided to the mounted `<FluentValidationValidator>` component (eg RuleSets) is ignored. The RuleSet will be used for whole-model validation, but ignored for field-level validation, as the configuration is not passed to the validator. At the field-level, rules are executed purely based on the property name, ignoring other configuration.

**To Reproduce**
I've put together a small Blazor Server app to demonstrate the issues: https://github.com/matthetherington/BlazoredFluentValidationIssueDemo

**Expected behavior**
* The same model-level validation rules apply to field-level validation, but scoped to the changed property.
* If options are passed to `<FluentValidationValidator>`, that configuration is used to validate both the whole model and individual fields
* If a Validator is provided to `<FluentValidationValidator>`, it is possible to validate sub object fields without an exception being thrown
* If a Validator is not provided, the type of `EditContext.Model` is used to locate a validator instead of `FieldIdentifier.Model`, so that rules defined on the root-level validator are ran.

In the demo app, I've put together a proposal for how it could behave - pages suffixed with "Fixed" use a modified version of Blazored.FluentValidation with the changes in #205

**Screenshots**
N/A

**Hosting Model (is this issue happening with a certain hosting model?):**
 N/A

### PR DESCRIPTION
**Pull request by matthetherington**
This fixes a few issues that are causing inconsistent behaviour between field-level and whole-model validation.

See #204 for more context and a demo project

Summary of changes:

* The type of `EditContext.Model` is used to locate a validator for field-level validation if one is not provided
* For field-level validation, the received `FieldIdentifier` is converted to a Fluent property path (eg `Customer.Address[0].Line1`) by searching the `EditContext.Model` instance, and then the validation results at that path are used to determine if the field is valid. The `<FluentValidationValidator>` options are supplied to Fluent, but are combined with an additional restriction so that only rules affecting the changed property are ran (as per the default `MemberNameValidatorSelector` behaviour)
* The configuration of `<FluentValidationValidator>` that is currently used for whole-model validation is also used when validating fields, this means that RuleSets and other configuration work as expected at the field-level.

The above would just mirror the existing whole-model validation for field-level validation, so I think it's a fairly safe change. If a project is using Blazored.FluentValidation then they'll have a validator configured that works for the whole model.

Could I please get your thoughts? I'm looking at this through quite a narrow lense so there may be issues with the proposal that haven't occured to me - it does seem to work okay though.

Resolves #204 
Resolves #188 
Resolves #180

### CODE REVIEW

**Comment by matthetherington** [comment]
Closing this while I make a few changes

**Comment by UniMichael** [comment]
Was asked to provide some feedback in #180, since we ended-up making our own validation handler at work to tackle something similar. 

From playing around with it, it does seem to properly validate nested objects' fields, though it's worth pointing-out that it doesn't address the problem discussed in #180 (all RuleSets firing, regardless of options), but it's like I said in that discussion: 

FluentValidation's API doesn't expose the necessary functions to fix that (which is why our fix at work validates the entire model on field change and only tracks the *new* validation errors so we can fake a field-speicific validation).

---

If anyone's interested in trying to fix that specific issue with this PR, here's the basic example I was using to reproduce it:

```cs
public class Person
{
    public string Name { get; set; } = string.Empty;
    public int Age { get; set; }
    public Address Address { get; set; } = new();
}

public class Address
{
    public string Street { get; set; } = string.Empty;
    public string Country { get; set; } = string.Empty;
    public string PostalCode { get; set; } = string.Empty;
}

public class PersonValidator : AbstractValidator<Person>
{
    public PersonValidator(IValidator<Address> addressValidator)
    {
        RuleFor(a => a.Address).SetValidator(addressValidator);
        
        RuleFor(p => p.Name).NotEmpty().WithMessage("Name required");
        
        RuleSet("Adult", () =>
        {
            RuleFor(p => p.Age).GreaterThan(18).WithMessage("Must be an adult");
        });
    }
}

public class AddressValidator : AbstractValidator<Address>
{
    public const string PostalCodePattern = "[a-zA-Z][0-0][a-zA-Z] [0-9][a-zA-Z][0-9]";
    public static readonly Regex PostalCodeRegex = new(PostalCodePattern);
    
    public AddressValidator()
    {
        RuleFor(a => a.Country).NotEmpty().WithMessage("Country required");
        RuleFor(a => a.Street).NotEmpty().WithMessage("Street required");
        RuleFor(a => a.PostalCode).Matches(PostalCodeRegex).WithMessage("Must be valid postal code");
        
        RuleSet("Canadian", () =>
        {
            RuleFor(a => a.Country)
                .NotEmpty().WithMessage("Country required")
                .Equal("Canada").WithMessage("Must be Canadian");
        });
    }
}
```

```razor
@page "/"
@using Microsoft.AspNetCore.Components.Forms
@using BlazorServerApp.Models
@using Blazored.FluentValidation

<EditForm Model="_person" OnValidSubmit="() => { }">
    @* NOTE: The options' rulesets are basically ignored, you can use whatever RuleSet you want, and it'll still run all of them *@
    <FluentValidationValidator Options="@(options => options.IncludeRuleSets("default"))"/>
    <ValidationSummary/>

    <h1>Person</h1>
    <div>
        <label>Name</label>
        <InputText @bind-Value="_person.Name"/>
    </div>
    <div>
        <label>Age</label>
        <InputNumber @bind-Value="_person.Age"/>
    </div>

    <h2>Address</h2>
    <div>
        <label>Country</label>
        <InputText @bind-Value="_person.Address.Country"/>
    </div>
    <div>
        <label>Street</label>
        <InputText @bind-Value="_person.Address.Street"/>
    </div>
    <div>
        <label>Postal Code</label>
        <InputText @bind-Value="_person.Address.PostalCode"/>
    </div>
    
    <button type="submit">
        Submit
    </button>
</EditForm>

@code {

    private readonly Person _person = new();

}
```

**Comment by matthetherington** [comment]
Thanks for that @UniMichael 

I tried to recreate the problem you were seeing, but it seems to be working okay when using the modified Blazored.FluentValidation from this PR.

Screenshot 2023-12-07 at 15 59 59

I can enter 17 for age, and UK as the country, and both fields are valid when blurring out.

Submitting the form also does not produce an validation message, so it does seem to be using only the default rule set and not including Adult / Canadian.

Screenshot 2023-12-07 at 16 01 10

Is that what you were expecting to happen?

**Comment by matthetherington** [comment]
It just occurred to me that if no `Options` are provided to `<FluentValidationValidator>:

* When validating the whole model, the default validator selector factory is used, which means only the default rule set will be executed
* When validating at the field level, because MemberNameValidatorSelector is used, the default validator selector factory is not used, and all rulesets are executed.

The exepected behaviour is decribed in the docs over at https://docs.fluentvalidation.net/en/latest/rulesets.html

> If you call Validate without passing a ruleset then only rules not in a RuleSet will be executed.

So I've just pushed a commit that will bring the field-level validation in-line with the model-level in that regard too.

**Comment by UniMichael** [comment]
@matthetherington Ok, I think I messed something up when I tried-out the PR. I may have forgotten to actually checkout the branch after I pulled it. I'm running it now and it does seem like the rule problem was fixed.

That's what I get for testing it on a Friday morning without having coffee, I guess. 😅

I'll edit my other comment.

**Comment by matthetherington** [comment]
@UniMichael ah brilliant, thanks for taking another look!

**Comment by Albiniivari** [comment]
@matthetherington Hey, so I've tried the PR and it seems to be working fine. However, I find the partial-validation a bit odd.
If you first input some non-allowed value in to the other fields, it displays a few errors. But if you then press the partial-validate, the existing errors are deleted, so that it seems that they are now OK. I would expect the partial-validate to validate only the fields included in rule-set, but leave existing errors etc to be?

File: src/Blazored.FluentValidation/PropertyPathHelper.cs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        {
+            var currentNode = nodes.Pop();
+            var currentModelObject = currentNode.ModelObject;
+
+            if (currentModelObject == fieldIdentifier.Model)
+            {
+                return BuildPropertyPath(currentNode, fieldIdentifier);
+            }
+            
+            var nonPrimitiveProperties = currentModelObject?.GetType()
```
Comment (jafin) [inline src/Blazored.FluentValidation/PropertyPathHelper.cs:35]: I had an issue with this bit of code, in a model it kept adding DateTime types on the NodeStack in an endless loop, eventually ending in out of memory.  I think additional type guards are needed here for types such as `DateTime`.  This change to check for additional simple types fixed the issue with DateTime not being a primitive. 
 https://github.com/jafin/FluentValidation/commit/2de01ab11cce79bdc5d2fc21e76a25724d325213

**Comment by etarvt** [comment]
@matthetherington I've been using the Blazored.FluentValidation library and find it to be excellent, but I encountered issue #204  a week ago. I've tested the PR, and it seems to resolve the problem with the sub-objects for me.

Do you expect this PR to be merged anytime soon?

**Comment by Xavia1991** [comment]
Please merge this soon, I have been waiting for this since end of last year when we switched to .net8.

**Comment by psandler** [comment]
Any timeline on when this will be reviewed/merged?  I am also having this problem and this PR seems to resolve it.

**Comment by ckirker** [comment]
I am in need of this fix as well--as I'm blocked using FluentValidation without it.

**Comment by DaveVW-AWE** [comment]
Any updates on merging this in?  Been waiting a VERY long time for this and it's holding up our project.  I like to rely on and support 3rd party packages however this has taken far too long to resolve.

**Comment by pwelter34** [comment]
apologies, I'll work to get this resolved and merged

**Review by pwelter34 — APPROVED** [review]

**Comment by dmorawetz** [comment]
Has anyone tested this PR with more than 5 fields and a non-trivial model?

We have hundreds of fields in our model with lists of more complex models. Because our form is user-configurable, we have effectively a tree of those complex models. The performance of `PropertyPathHelper.ToFluentPropertyPath` is making this package unusable for us.

We use this package since 2021 and had never issues with the default `MemberNameValidatorSelector`.

Concerning the original issue from #204, I think there should be a better way to select the validator of the root model, if needed. It should at least be configurable, as for our project, the previous selection strategy worked better.

**Pull request merged**